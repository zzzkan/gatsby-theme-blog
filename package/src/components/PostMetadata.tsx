import React from "react"
import { Box, Icon, HStack, Wrap, WrapItem } from "@chakra-ui/react"
import { BiCalendar } from "@react-icons/all-files/bi/BiCalendar"
import { BiCalendarEdit } from "@react-icons/all-files/bi/BiCalendarEdit"
import { BiTime } from "@react-icons/all-files/bi/BiTime"

export type PostMetadataProps = {
  readonly publishedDate: string
  readonly updatedDate: string | null
  readonly publishedDate_ISO8601: string
  readonly updatedDate_ISO8601: string | null
  readonly timeToReadMinutes: number | null
}

const PostMetadata: React.FC<PostMetadataProps> = (post) => {
  return (
    <Wrap color={"base.500"} fontSize={"sm"} spacingX={1} spacingY={0}>
      <WrapItem>
        <HStack spacing={0}>
          <Icon as={BiCalendar} fontSize={"md"} marginTop={-0.5} />
          <Box as={"time"} dateTime={post.publishedDate_ISO8601}>
            {post.publishedDate}
          </Box>
        </HStack>
      </WrapItem>
      {post.updatedDate != null && post.updatedDate_ISO8601 != null && (
        <WrapItem>
          <HStack spacing={0}>
            <Icon as={BiCalendarEdit} fontSize={"md"} marginTop={-0.5} />
            <Box as={"time"} dateTime={post.updatedDate_ISO8601}>
              {post.updatedDate}
            </Box>
          </HStack>
        </WrapItem>
      )}
      {post.timeToReadMinutes != null && (
        <WrapItem>
          <HStack spacing={0}>
            <Icon as={BiTime} fontSize={"md"} marginTop={-0.5} />
            <Box as={"time"}>{Math.ceil(post.timeToReadMinutes)}min</Box>
          </HStack>
        </WrapItem>
      )}
    </Wrap>
  )
}

export default PostMetadata
