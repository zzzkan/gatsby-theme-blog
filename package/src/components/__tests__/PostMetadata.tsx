import React from "react"
import { render } from "@testing-library/react"
import { PostMetadata } from "../PostMetadata"

describe("PostMetadata component", () => {
  test("snapshot", () => {
    const { asFragment } = render(
      <PostMetadata
        publishedDate={"2022/01/01"}
        updatedDate={"2022/01/02"}
        publishedDateISO8601={"2022-01-01"}
        updatedDateISO8601={"2022-01-02"}
        timeToReadMinutes={10}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  test("should render published date as text content, published date (ISO8601) as datetime attribute", () => {
    const { queryByText } = render(
      <PostMetadata
        publishedDate={"2022/01/01"}
        updatedDate={null}
        publishedDateISO8601={"2022-01-01"}
        updatedDateISO8601={null}
        timeToReadMinutes={null}
      />
    )
    expect(queryByText("2022/01/01")).toHaveAttribute("datetime", "2022-01-01")
    expect(queryByText("2022-01-01")).toBeNull()
  })

  test("should not render updated date", () => {
    const { queryByText } = render(
      <PostMetadata
        publishedDate={"2022/01/01"}
        updatedDate={"2022/01/02"}
        publishedDateISO8601={"2022-01-01"}
        updatedDateISO8601={null}
        timeToReadMinutes={null}
      />
    )
    expect(queryByText("2022/01/02")).toBeNull()
  })

  test("should not render updated date (ISO8601)", () => {
    const { queryByText } = render(
      <PostMetadata
        publishedDate={"2022/01/01"}
        updatedDate={null}
        publishedDateISO8601={"2022-01-01"}
        updatedDateISO8601={"2022-01-02"}
        timeToReadMinutes={null}
      />
    )
    expect(queryByText("2022-01-02")).toBeNull()
  })

  test("should render updated date as text content, updated date (ISO8601) as datetime attribute", () => {
    const { queryByText } = render(
      <PostMetadata
        publishedDate={"2022/01/01"}
        updatedDate={"2022/01/02"}
        publishedDateISO8601={"2022-01-01"}
        updatedDateISO8601={"2022-01-02"}
        timeToReadMinutes={null}
      />
    )
    expect(queryByText("2022/01/02")).toHaveAttribute("datetime", "2022-01-02")
    expect(queryByText("2022-01-02")).toBeNull()
  })

  test("should render time to read on text content", () => {
    const { queryByText } = render(
      <PostMetadata
        publishedDate={"2022/01/01"}
        updatedDate={"2022/01/02"}
        publishedDateISO8601={"2022-01-01"}
        updatedDateISO8601={"2022-01-02"}
        timeToReadMinutes={99}
      />
    )
    expect(queryByText("99min")).toBeInTheDocument()
  })
})
