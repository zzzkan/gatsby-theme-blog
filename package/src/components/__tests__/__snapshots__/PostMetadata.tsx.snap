// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PostMetadata component > snapshot 1`] = `
<DocumentFragment>
  .emotion-0 {
  overflow: hidden;
  color: secondaryText;
  font-size: sm;
}

.emotion-1 {
  --chakra-wrap-x-spacing: 1px;
  --chakra-wrap-y-spacing: 0px;
  --wrap-x-spacing: calc(var(--chakra-wrap-x-spacing) / 2);
  --wrap-y-spacing: calc(var(--chakra-wrap-y-spacing) / 2);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  list-style-type: none;
  padding: 0px;
  margin: calc(var(--wrap-y-spacing) * -1) calc(var(--wrap-x-spacing) * -1);
}

.emotion-1>*:not(style) {
  margin: var(--wrap-y-spacing) var(--wrap-x-spacing);
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.emotion-3>*:not(style)~*:not(style) {
  margin-top: 0px;
  -webkit-margin-end: 0px;
  margin-inline-end: 0px;
  margin-bottom: 0px;
  -webkit-margin-start: 0px;
  margin-inline-start: 0px;
}

.emotion-4 {
  width: 1em;
  height: 1em;
  display: inline-block;
  line-height: 1em;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  color: currentColor;
  font-size: md;
}

<div
    class="chakra-wrap emotion-0"
  >
    <ul
      class="chakra-wrap__list emotion-1"
    >
      <li
        class="chakra-wrap__listitem emotion-2"
      >
        <div
          class="chakra-stack emotion-3"
        >
          <svg
            class="chakra-icon emotion-4"
            fill="currentColor"
            focusable="false"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7 11H9V13H7zM7 15H9V17H7zM11 11H13V13H11zM11 15H13V17H11zM15 11H17V13H15zM15 15H17V17H15z"
            />
            <path
              d="M5,22h14c1.103,0,2-0.897,2-2V8V6c0-1.103-0.897-2-2-2h-2V2h-2v2H9V2H7v2H5C3.897,4,3,4.897,3,6v2v12 C3,21.103,3.897,22,5,22z M19,8l0.001,12H5V8H19z"
            />
          </svg>
          <time
            class="emotion-5"
            datetime="2022-01-01"
          >
            2022/01/01
          </time>
        </div>
      </li>
      <li
        class="chakra-wrap__listitem emotion-2"
      >
        <div
          class="chakra-stack emotion-3"
        >
          <svg
            class="chakra-icon emotion-4"
            fill="currentColor"
            focusable="false"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19,4h-3V2h-2v2h-4V2H8v2H5C3.897,4,3,4.897,3,6v14c0,1.103,0.897,2,2,2h14c1.103,0,2-0.897,2-2V6C21,4.897,20.103,4,19,4z M5,20V7h3h2h4h2h3V6l0.002,14H5z"
            />
            <path
              d="M14.142 9.628H16.686999999999998V11.567H14.142z"
              transform="rotate(-135.016 15.414 10.598)"
            />
            <path
              d="M8.005 16.201L8.005 18 9.804 18 14.78 13.03 12.981 11.231z"
            />
          </svg>
          <time
            class="emotion-5"
            datetime="2022-01-02"
          >
            2022/01/02
          </time>
        </div>
      </li>
      <li
        class="chakra-wrap__listitem emotion-2"
      >
        <div
          class="chakra-stack emotion-3"
        >
          <svg
            class="chakra-icon emotion-4"
            fill="currentColor"
            focusable="false"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12,2C6.486,2,2,6.486,2,12s4.486,10,10,10c5.514,0,10-4.486,10-10S17.514,2,12,2z M12,20c-4.411,0-8-3.589-8-8 s3.589-8,8-8s8,3.589,8,8S16.411,20,12,20z"
            />
            <path
              d="M13 7L11 7 11 13 17 13 17 11 13 11z"
            />
          </svg>
          <span
            class="emotion-5"
          >
            10min
          </span>
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
